sql/src/test/java/io/crate/operation/projectors/FetchProjectorTest.java

```diff
@@ -0,0 +1,190 @@
+        verify(upstream, timeout(100).times(3)).resume(false);
+        fetchProjector.finish();
+
+        Bucket projected = rowReceiver.result();
+        assertThat(projected.size(), is(10));
+
+        int iterateLength = Iterables.size(rowReceiver.result());
+        assertThat(iterateLength, is(10));
+    }
+
+
+    private FetchProjector prepareFetchProjector(int fetchSize,
+                                                 CollectingRowReceiver rowReceiver,
+                                                 TransportFetchOperation transportFetchOperation,
+                                                 RowUpstream upstream) {
+        FetchProjector pipe =
+            new FetchProjector(
+                transportFetchOperation,
+                threadPool.executor(ThreadPool.Names.SUGGEST),
+                TestingHelpers.getFunctions(),
+                buildOutputSymbols(),
+                buildFetchProjectorContext(),
+                fetchSize
+            );
+        pipe.setUpstream(upstream);
+        pipe.downstream(rowReceiver);
+        pipe.prepare();
+        return pipe;
+    }
+
+    private FetchProjectorContext buildFetchProjectorContext() {
+        Map<String, IntSet> nodeToReaderIds = new HashMap<>(2);
+        IntSet nodeReadersNodeOne = new IntHashSet();
+        nodeReadersNodeOne.add(0);
+        IntSet nodeReadersNodeTwo = new IntHashSet();
+        nodeReadersNodeTwo.add(2);
+        nodeToReaderIds.put("nodeOne", nodeReadersNodeOne);
+        nodeToReaderIds.put("nodeTwo", nodeReadersNodeTwo);
+
+        TreeMap<Integer, String> readerIndices = new TreeMap<>();
+        readerIndices.put(0, "t1");
+
+        Map<String, TableIdent> indexToTable = new HashMap<>(1);
+        indexToTable.put("t1", USER_TABLE_IDENT);
+
+        ReferenceIdent referenceIdent = new ReferenceIdent(USER_TABLE_IDENT, "id");
+        ReferenceInfo referenceInfo = new ReferenceInfo(referenceIdent,
+            RowGranularity.DOC,
+            LongType.INSTANCE,
+            ColumnPolicy.STRICT,
+            ReferenceInfo.IndexType.NOT_ANALYZED);
+
+        Map<TableIdent, FetchSource> tableToFetchSource = new HashMap<>(2);
+        FetchSource fetchSource = new FetchSource(Collections.<ReferenceInfo>emptyList(),
+            Collections.singletonList(new InputColumn(0)),
+            Collections.singletonList(new Reference(referenceInfo))
+            );
+        tableToFetchSource.put(USER_TABLE_IDENT, fetchSource);
+
+        return new FetchProjectorContext(
+            tableToFetchSource,
+            nodeToReaderIds,
+            readerIndices,
+            indexToTable
+        );
+    }
+
+    private List<Symbol> buildOutputSymbols() {
+        List<Symbol> outputSymbols = new ArrayList<>(2);
+
+        InputColumn inputColumn = new InputColumn(0);
+        ReferenceIdent referenceIdent = new ReferenceIdent(USER_TABLE_IDENT, "id");
+        ReferenceInfo referenceInfo = new ReferenceInfo(referenceIdent,
```

Comment:
Just FYI: To create Reference's there would also be `SqlExpressions.asSymbol(...)` - but I'm not sure if it makes sense to use it here